src/uct/api/v2/uct_v2.h

```diff
@@ -948,7 +948,13 @@ typedef enum {
      * packed key by @ref uct_md_mkey_pack_v2 with
      * @ref UCT_MD_MKEY_PACK_FLAG_INVALIDATE flag.
      */
-    UCT_MD_FLAG_INVALIDATE_AMO = UCS_BIT(12)
+    UCT_MD_FLAG_INVALIDATE_AMO = UCS_BIT(12),
+
+    /**
+     * Memory domain supports remote flush using @ref uct_ep_flush or @ref
+     * uct_iface_flush with flag @ref UCT_FLUSH_FLAG_REMOTE.
+     */
+    UCT_MD_FLAG_FLUSH_REMOTE   = UCS_BIT(13)
```

Comment:
can we avoid API change in this PR?

Reply:
I see following possiblities:
1. keep `HAVE_MLX5_DV` as is (it's not worse than the current check on `rc_verbs` regardless of the MD tbh)
2. split that API part adding the flag in a separate PR
3. keep `HAVE_MLX5_DV` and add some IB ops pointer checks on `md()->ops`
4. Try to add function implementing 3. in some private UCT headers and use it here

Suggest to keep 1. and start PR for 2. and the last PR of the two that gets merged with resolve conflict.

Reply:
can we check if md->ops->close points to uct_ib_mlx5_devx_md_close?

Reply:
We can but only under `HAVE_MLX5_DV`. Otherwise, it does not work when gtest is not build with libuct_ib or libuct_ib_mlx5 (i think I saw that on CI cuda docker test). Is that what you meant?

Reply:
yes. the problem with checking only HAVE_MLX5_DV, without md->ops, is that we could build with mlx5 but create md without devx (if, for example, devx is disabled in fw)

Reply:
tried `((GetParam()->tl_name != "rc_verbs") || sender().md()->ops->close != uct_ib_mlx5_devx_md_close)` but fails link time with missing symbol as we link with .so. (tried with other ib only functions, same issue). Trying with casting to ib md and checking name.

Reply:
BTW, which test variants fail if we don't have any skip check?
For example, I'd expect all non-RDMA transports to work, since there is to PCI sync issue.

Reply:
I'd suggest the following patch:
```diff
diff --git a/test/gtest/uct/test_flush.cc b/test/gtest/uct/test_flush.cc
index e222b50f57..ec44eef944 100644
--- a/test/gtest/uct/test_flush.cc
+++ b/test/gtest/uct/test_flush.cc
@@ -12,19 +12,7 @@ extern "C" {

 #if HAVE_MLX5_DV
 #include <uct/ib/base/ib_md.h>
-
-static bool tl_is_rc_verbs_devx(std::string tl_name, uct_md_h md)
-{
-    // Only MD created with MLX5 provider supports remote flush
-    return tl_name == "rc_verbs" &&
-        std::string(((uct_ib_md_t *)md)->name) == "ib_mlx5";
-}
-#else
-static bool tl_is_rc_verbs_devx(std::string tl_str GTEST_ATTRIBUTE_UNUSED_,
-                                uct_md_h md GTEST_ATTRIBUTE_UNUSED_)
-{
-    return false;
-}
+#include <uct/ib/mlx5/ib_mlx5.h>
 #endif

 class uct_flush_test : public uct_test {
@@ -165,11 +153,18 @@ public:
     }

     void check_skip_test_flush_remote() {
-        if ((GetParam()->tl_name != "dc_mlx5") &&
-            !tl_is_rc_verbs_devx(GetParam()->tl_name, sender().md()) &&
-            (GetParam()->tl_name != "rc_mlx5")) {
-            UCS_TEST_SKIP_R("not supported");
+#ifdef HAVE_MLX5_DV
+        auto md = sender().md();
+        if (std::string(md->component->name) == "ib") {
+            auto ib_md = ucs_derived_of(md, uct_ib_md_t);
+            if (ib_md->dev.flags & UCT_IB_DEVICE_FLAG_MLX5_PRM) {
+                auto mlx5_md = ucs_derived_of(ib_md, uct_ib_mlx5_md_t);
+                if (!(mlx5_md->flags & UCT_IB_MLX5_MD_FLAG_KSM)) {
+                    UCS_TEST_SKIP_R("not supported");
+                }
+            }
         }
+#endif

         m_flush_flags = UCT_FLUSH_FLAG_REMOTE;
     }
@@ -509,8 +504,8 @@ UCS_TEST_SKIP_COND_P(uct_flush_test, put_bcopy_flush_ep_remote,
 }

 UCS_TEST_SKIP_COND_P(uct_flush_test, put_zcopy_flush_ep_remote,
-                     !check_caps(UCT_IFACE_FLAG_PUT_BCOPY) ||
-                     !check_caps(UCT_IFACE_FLAG_GET_BCOPY)) {
+                     !check_caps(UCT_IFACE_FLAG_PUT_ZCOPY) ||
+                     !check_caps(UCT_IFACE_FLAG_GET_ZCOPY)) {
     check_skip_test_flush_remote();
     test_flush_put_zcopy(&uct_flush_test::flush_ep_nb);
 }
```

Reply:
fixed, failing are below:
```
[  FAILED  ] rc_verbs/uct_flush_test.put_bcopy_flush_ep_remote/0, where GetParam() = rc_verbs/mlx5_0:1
[  FAILED  ] rc_verbs/uct_flush_test.put_bcopy_flush_ep_remote/1, where GetParam() = rc_verbs/mlx5_1:1
[  FAILED  ] rc_verbs/uct_flush_test.put_bcopy_flush_ep_remote/2, where GetParam() = rc_verbs/mlx5_2:1
[  FAILED  ] rc_verbs/uct_flush_test.put_zcopy_flush_ep_remote/0, where GetParam() = rc_verbs/mlx5_0:1
[  FAILED  ] rc_verbs/uct_flush_test.put_zcopy_flush_ep_remote/1, where GetParam() = rc_verbs/mlx5_1:1
[  FAILED  ] rc_verbs/uct_flush_test.put_zcopy_flush_ep_remote/2, where GetParam() = rc_verbs/mlx5_2:1
```

with

```
[ RUN      ] rc_verbs/uct_flush_test.put_zcopy_flush_ep_remote/0 <rc_verbs/mlx5_0:1>
[1721026779.013664] [rock03:11330:0]     rc_verbs_ep.c:488  UCX  ERROR rcv endpoint 0x3795e90 on mlx5_0: flush(remote) is not supported
uct/test_flush.cc:333: Failure
Error: Unsupported operation
[1721026779.016897] [rock03:11330:0]           rc_ep.c:488  UCX  WARN  destroying txqp 0x3795ea0 with uncompleted operation 0x384f010 handler uct_rc_ep_send_op_completion_handler
common/test.cc:366: Failure
Failed
Got 1 errors and 1 warnings during the test
[     INFO ] < rc/verbs/rc_verbs_ep.c:488 rcv endpoint 0x3795e90 on mlx5_0: flush(remote) is not supported >
[     INFO ] < rc/base/rc_ep.c:488 destroying txqp 0x3795ea0 with uncompleted operation 0x384f010 handler uct_rc_ep_send_op_completion_handler >
[  FAILED  ] rc_verbs/uct_flush_test.put_zcopy_flush_ep_remote/0, where GetParam() = rc_verbs/mlx5_0:1 (46 ms)
```

Reply:
We can also skip rc_verbs explicitly

Reply:
fixed

Reply:
reverted to suggested patch because i think i did not understand your latest suggestion. we cannot skip rc_verbs explicitly because it might actually support remote flush when created with devx md.